pbft/pbft.cpp

```diff
@@ -994,6 +1007,533 @@ pbft::get_high_water_mark()
+                    , this->prepared_operations_since_last_checkpoint())
+    };
+
+    this->view_is_valid = false;
+    this->last_view_sent = msg.view();
+    this->broadcast(this->wrap_message(viewchange_message));
+
+}
+
+void
+pbft::handle_viewchange(const pbft_msg& msg, const bzn_envelope& original_msg)
+{
+    if (this->is_valid_viewchange_message(msg, original_msg))
+    {
+        this->valid_view_change_messages[msg.view()].insert(original_msg.SerializeAsString());
+    }
+    else
+    {
+        LOG(error) << "handle_viewchange - invalid viewchange message, ignoring";
+        return;
+    }
+
+    this->save_checkpoint(msg);
+
+    // When a replica recieves f + 1 view change messages it sends one as well
+    // should it be == or >=  ?
+    if (this->view_is_valid && this->valid_view_change_messages.size() == this->max_faulty_nodes() + 1 && msg.view() > this->last_view_sent)
+    {
+        this->replica_broadcasts_viewchange(msg);
+    }
+
+    // TODO move this to the end...
+    // we want to filter std::map<uint64_t, std::set<bzn_envelope>> valid_view_change_messages for
+    // 1) we would be the primary for that view
+    // 2) that view is greater than the current view
+    // 3) we have 2 f + 1 viewchange messages for it
+    const auto viewchange = std::find_if(this->valid_view_change_messages.begin(), this->valid_view_change_messages.end(), [&](const auto& p)
+            {
+                return ((this->get_primary(p.first).uuid == this->get_uuid()) &&  (p.first > this->view) && (p.second.size() >= 2 * this->max_faulty_nodes() + 1));
+            });
+
+    if (viewchange == this->valid_view_change_messages.end())
+    {
+        LOG (debug) << "handle_viewchange - not enough viewchange messages";
+        return;
+    }
+
+    // When the primary of the new view recieves 2f valid view change messages for the new view
+    // the primary of new view broadcasts NEWVIEW
+
+    // TODO Refactor into method
+    {
+        // create the newview and and broadcast it
+
+        // viewchange->second is a set of bzn_envelope strings
+        // -- we want a of set<pbft_msg>
+
+        std::vector<pbft_msg> viewchange_messages;
+        for (const auto& view_change_msg : viewchange->second)
+        {
+            pbft_msg pbft_viewchange;
+            if (!pbft_viewchange.ParseFromString(view_change_msg))
+            {
+                LOG (error) << "handle_viewchange - unable to parse viewchange message";
+                continue;
+            }
+
+            // Since sets require hashable objects we needed to use vectors and
+            // do our own uniqueness testing.
+            // TODO: use a set with a hash function
+            auto found_iter = std::find_if(std::cbegin(viewchange_messages) , std::cend(viewchange_messages) , [&](const auto& viewchange)
+            {
+                return !google::protobuf::util::MessageDifferencer::Equals(viewchange, pbft_viewchange);
```

Comment:
I don't trust MessageDifferencer here, given the things we're doing with serialized messages.

Reply:
I  have refactored the code to remove the reliance on MessageDifferencer.